### DISCUSSION TITLE
Why docs recommend useState for routes?

### DISCUSSION THREAD

**Discussion by slavko-lukic** [react-navigation/react-navigation#11905] [Q&A] [answered]
TabView docs show us that we need to declare routes this way with useState:
```javascript
  const [routes] = React.useState([
    { key: 'first', title: 'First' },
    { key: 'second', title: 'Second' },
  ]);
  
    return (
    <TabView
      navigationState={{ index, routes }}
      renderScene={renderScene}
      onIndexChange={setIndex}
      initialLayout={{ width: layout.width }}
    />
  );
```

Is there a reason for that in regards with optimization?

I'm asking because I have variable in the tab label so I have to use memo for this

```javascript
  const routes = React.useMemo(
    () => [
      { key: "first", title: `First ${someLabel}` },
      { key: "second", title: "Second" },
    ],
    [someLabel],
  );
  
    return (
    <TabView
      navigationState={{ index, routes }}
      renderScene={renderScene}
      onIndexChange={setIndex}
      initialLayout={{ width: layout.width }}
    />
  );
```

**Comment by satya164** [comment]
With the `useMemo`, all route objects will be recreated when `someLabel` changes - re-rendering all the tabs which can be a perf issue. With `useState`, you can update only the route that changed and only the tab with that route will re-render.

**Reply by slavko-lukic** [reply]
@satya164 
So you are suggesting something like this?

```javascript
  const firstRoute = useMemo(
    () => ({ key: "first", title: someLabel }),
    [someLabel],
  );

  const [secondRoute] = useState({ key: "second", title: "Second" });

  return (
    <TabView
      navigationState={{ index, routes: [firstRoute, secondRoute] }}
      renderScene={renderScene}
      onIndexChange={setIndex}
      initialLayout={{ width: layout.width }}
    />
```

**Reply by satya164** [reply, marked as answer]
My comment was about if you're updating the state later, but if you need to derive the route from another prop/state, yeah that's fine.

The main thing is to avoid re-creating the `route` objects unless something changes in the route.

**Reply by satya164** [reply]
And if your route objects never change, you can just define them outside of the component. No need for state.
